test/controllers/tasks_controller_test.rb

```diff
@@ -77,52 +77,170 @@
-      expect(new_task.due_date.to_time.to_i).must_equal task_hash[:task][:due_date].to_i
       expect(new_task.completed).must_equal task_hash[:task][:completed]
 
       must_respond_with :redirect
-      must_redirect_to task_path(new_task.id)
+      must_redirect_to tasks_path
+    end
+
+    # THE BELOW TEST IS GIVING AN ERROR AS FOLLOWS
+    # ActionController::ParameterMissing:
+    # param is missing or the value is empty: task
+    # THIS IS A TEST I ADDED AND IS NOT REQUIRED. INSPIRATION CAME FROM
+    # https://github.com/Ada-Developers-Academy/textbook-curriculum/blob/master/08-rails/code_samples/create_controller_test.rb
+
+    it "will not create a task with invalid params" do
+      skip
+      task_hash = {} # invalid params
+
+      expect {
+        post tasks_path, params: task_hash
+      }.wont_change "Task.count"
+
+      must_respond_with :bad_request
     end
   end
 
   # Unskip and complete these tests for Wave 3
   describe "edit" do
     it "can get the edit page for an existing task" do
-      skip
-      # Your code here
+      # skip
+      # Act
+      get edit_task_path(task.id)
+
+      # Assert
+      must_respond_with :success
     end
 
     it "will respond with redirect when attempting to edit a nonexistant task" do
-      skip
-      # Your code here
+      # skip
+      get edit_task_path(-1)
+
+      # Assert
+      must_respond_with :redirect
     end
   end
 
   # Uncomment and complete these tests for Wave 3
   describe "update" do
-    # Note:  If there was a way to fail to save the changes to a task, that would be a great
-    #        thing to test.
+    # Note:  If there was a way to fail to save the changes to a task, that would be a great thing to test.
+
+    before do
+      Task.create(task_name: "Clean the bathroom", description: "All of it! Don't forget the floor.", date_due: "2020-01-01")
+    end
+
+    let (:new_task_hash) {
+      {
+        task: {
+          task_name: "Buy groceries",
+          description: "Silk vanilla almond creamer and bacon",
+          date_due: "2019-04-15".to_date,
+          completed: false,
+        },
+      }
+    }
+
+    it "can update an existing task" do
+      # skip
+
+      id = Task.last.id
+      expect {
+        patch task_path(id), params: new_task_hash
+      }.wont_change "Task.count"
+
+      must_respond_with :redirect
+
+      task = Task.find_by(id: id)
+      expect(task.task_name).must_equal new_task_hash[:task][:task_name]
+      expect(task.description).must_equal new_task_hash[:task][:description]
+      expect(task.date_due).must_equal new_task_hash[:task][:date_due]
+    end
+
+    # THIS IS THE FIRST THING I TRIED WRITING. IT GAVE AN ERROR.
+    # ActiveModel::UnknownAttributeError: unknown attribute 'params' for Task.
     it "can update an existing task" do
       skip
-      # Your code here
+      # Arrange
+      change_task = Task.find_by(task_name: "Clean the bathroom")
+
+      change_task.update(params: new_task_hash)
```

Comment:
Instead of calling `update` on the task itself, you'll want to use routes to get to the controller action you are testing. The method `update` that's on the task itself doesn't take a `params` hash... but the controller update action does! That's why your other version makes sense-- you call `patch task_path(id), params: new_task_hash`